src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java

```diff
@@ -0,0 +1,671 @@
+import accord.impl.progresslog.TxnStateKind;
+import accord.local.CommandStores;
+import accord.local.DurableBefore;
+import accord.local.MaxConflicts;
+import accord.local.Node;
+import accord.local.RedundantBefore;
+import accord.local.RejectBefore;
+import accord.primitives.Routable;
+import accord.primitives.Status;
+import accord.primitives.Timestamp;
+import accord.primitives.Txn;
+import accord.primitives.TxnId;
+import accord.utils.Invariants;
+import accord.utils.async.AsyncChain;
+import org.apache.cassandra.cql3.statements.schema.CreateTableStatement;
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.db.marshal.ByteBufferAccessor;
+import org.apache.cassandra.db.marshal.Int32Type;
+import org.apache.cassandra.db.marshal.LongType;
+import org.apache.cassandra.db.marshal.TupleType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.db.marshal.UUIDType;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.service.accord.AccordCommandStore;
+import org.apache.cassandra.service.accord.AccordKeyspace;
+import org.apache.cassandra.service.accord.AccordService;
+import org.apache.cassandra.service.accord.AccordStateCache;
+import org.apache.cassandra.service.accord.CommandStoreTxnBlockedGraph;
+import org.apache.cassandra.service.accord.api.AccordRoutingKey;
+import org.apache.cassandra.service.accord.api.AccordRoutingKey.TokenKey;
+import org.apache.cassandra.service.consensus.migration.ConsensusMigrationState;
+import org.apache.cassandra.service.consensus.migration.TableMigrationState;
+import org.apache.cassandra.tcm.ClusterMetadata;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.Pair;
+
+import static java.lang.String.format;
+import static java.util.Comparator.comparing;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static accord.utils.async.AsyncChains.getBlockingAndRethrow;
+import static org.apache.cassandra.schema.SchemaConstants.VIRTUAL_ACCORD_DEBUG;
+import static org.apache.cassandra.utils.ByteBufferUtil.bytes;
+import static org.apache.cassandra.utils.MonotonicClock.Global.approxTime;
+
+public class AccordDebugKeyspace extends VirtualKeyspace
+{
+    // {epoch, hlc, flags, node}
+    private static final TupleType TIMESTAMP_TYPE =
+        new TupleType(List.of(LongType.instance, LongType.instance, Int32Type.instance, Int32Type.instance));
+    private static final String TIMESTAMP_TYPE_STRING = TIMESTAMP_TYPE.asCQL3Type().toString();
+
+    // {epoch, hlc, kind, domain, node}
+    private static final TupleType TXN_ID_TYPE =
```

Comment:
In theory a `TxnId` can have other flags. This is not used at the moment, but we might want to future-proof this, and report the raw flags as well? An example might be the "coordinator optimisation" that permits smaller quorums at the expense of slightly more coordinator work, that we expect to implement before the new year (and for which I think the easiest way to configure it would be with a `TxnId` flag) - but no doubt there will be other flags we want to introduce later.

Reply:
Alternatively, reporting the `String` representation as suggested by David isn't a bad idea really, and removing all of the UDT/Tuple across the board for these types.

Reply:
Addressed by switching all timestamps and txnIds to `text`.